capa/features/extractors/ts/tools.py

```diff
@@ -0,0 +1,243 @@
+    integer_suffixes: Tuple[str, ...]  # Tends to indicate unsigned (100u) or long (100l) integer literal
+
+    def __init__(self):
+        self.import_signatures = self.load_import_signatures(self.signature_file)
+
+    def load_import_signatures(self, signature_file: str) -> Dict[str, set[str]]:
+        signatures = json.loads(importlib.resources.read_text(capa.features.extractors.ts.signatures, signature_file))
+        return {category: set(namespaces) for category, namespaces in signatures.items()}
+
+    def _is_import(self, name: str) -> bool:
+        return name in self.import_signatures["imports"]
+
+    def is_builtin(self, func: str) -> bool:
+        return func in self.import_signatures["builtins"]
+
+    def join_names(self, *args: str) -> str:
+        return ".".join(args)
+
+    def split_name(self, name: str) -> List[str]:
+        return name.split(".")
+
+    def process_property(self, node: Node, name: str) -> Optional[str]:
+        if self.is_method_call(node):  # yield only p.StartInfo but not p.Start()
+            return None
+        if self.is_recursive_property(node):  # yield only Current.Server.ClearError but not Current.Server and Current
+            return None
+        return self.join_names(*self.split_name(name)[1:])
+
+    def process_imported_constant(self, node: Node, name: str) -> Optional[str]:
+        if self.is_method_call(node):  # yield only ssl.CERT_NONE and not ssl.wrap_socket()
+            return None
+        if self.is_recursive_property(node):  # yield foo.foo.bar and not foo.bar or bar
+            return None
+        return name
+
+    def format_imported_class(self, name: str) -> str:
+        return name
+
+    def format_imported_class_members(self, name: str) -> str:
+        qualified_names = self.split_name(name)
+        if len(qualified_names) < 2:
+            raise ValueError(f"{name} does not have an associated class or namespace")
+        if len(qualified_names) == 2:
+            classname, membername = qualified_names[0], qualified_names[1]
+            return f"{classname}::{membername}"
+        namespace, classname, membername = qualified_names[:-2], qualified_names[-2], qualified_names[-1]
+        return f"{'.'.join(namespace)}.{classname}::{membername}"
+
+    def format_imported_function(self, name: str) -> str:
+        return self.format_imported_class_members(name)
+
+    def format_imported_property(self, name: str) -> str:
+        return self.format_imported_class_members(name)
+
+    def format_imported_constant(self, name: str) -> str:
+        return self.format_imported_class_members(name)
+
+    def parse_integer(self, integer: str) -> int:
+        for suffix in self.integer_suffixes:
+            if integer.endswith(suffix):
+                integer = integer[:-1]
+        for prefix, base in self.integer_prefixes:
+            if integer.startswith(prefix):
+                return int(integer, base)
+        return int(integer)
```

Comment:
do we expect exceptions to occur here?

Reply:
This can (and does raise ValueError when TS labels something as an int which is not an int), but the ValueError is handled by the caller (see extract_integers).